// @graphql-modules/di/dist/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DESIGN_PARAM_TYPES = 'design:paramtypes';
exports.DESIGN_TYPE = 'design:type';
exports.PROVIDER_OPTIONS = 'provider-options';
exports.PROPERTY_KEYS = 'property-keys';
function getServiceIdentifierName(serviceIdentifier) {
    if (typeof serviceIdentifier === 'function' && isType(serviceIdentifier)) {
        return serviceIdentifier.name;
    }
    else if (typeof serviceIdentifier !== 'undefined') {
        return serviceIdentifier.toString();
    }
    else {
        return 'undefined';
    }
}
exports.getServiceIdentifierName = getServiceIdentifierName;
function isType(v) {
    return typeof v === 'function' && 'prototype' in v;
}
exports.isType = isType;
function isTypeProvider(v) {
    return isType(v);
}
exports.isTypeProvider = isTypeProvider;
function isValueProvider(v) {
    return 'useValue' in v;
}
exports.isValueProvider = isValueProvider;
function isClassProvider(v) {
    return 'useClass' in v && isType(v.useClass);
}
exports.isClassProvider = isClassProvider;
function isFactoryProvider(v) {
    return 'useFactory' in v && typeof v.useFactory === 'function';
}
exports.isFactoryProvider = isFactoryProvider;
